accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,608 @@
+
+The `JsonSerializerOptions` is passed to `JsonNode` and used to specify any run-time added custom converters, handle other options such as quoted numbers, and to respect the `PropertyNameCaseInsensitive` to determine whether the string-based lookup of property names on `JsonObject` are case-sensitive (for both direct use of `JsonObject` and also indirect use through `dynamic`). If we have `static JsonValue Parse()` methods, those would also need an override to specify the `JsonSerializerOptions`.
+
+Although the serializer can be used to directly (de)serialize `JsonNode` types, once a `JsonNode` instance is created, it is expected the `JsonNode.GetValue<T>()` method is used to change types instead of using the serializer:
+```cs
+    JsonNode node = ... // any value
+    string str;
+
+    // This works, but has no caching:
+    byte[] utf8 = JsonSerializer.SerializeToUtf8Bytes(node, options);
+    str = JsonSerializer.Deserialize<string>(utf8, options);
+
+    // For ease-of-use and performance, use GetValue().
+    // It may return the raw internal value or the last converted value instead of deserializing.
+    str = node.GetValue<string>();
+```
+
+### Cast operators
+No explicit or implicit cast operators exist since `GetValue<T>` allows for a single, consistent programming model. This also works on not just the "known" types in the CLR but also custom data types known by the serializer.
+
+Cast operators do not work in all languages such as F# and even for C# would not work nicely since `System.Object` is allowed for `JsonArray` and `JsonCollection` elements. For example:
+```cs
+JsonArray jArray = ...
+jArray[2] = "hello";
+// Is jArray[2] a string or a JsonValue?
+
+jArray[2] = 2;
+// jArray[2] is always an int since we can't have implicit cast operators for number types.
+```
+
+Although it is possible to add cast operators for `JsonValue` to\from string\bool, number types should not have any implicit operator since they may throw - for example, an internal value of "3.14" cannot be converted into `int` so an exception must be thrown.
+
+### Specifying the `JsonSerializerOptions` instance
+The `JsonSerializerOptions` instance needs to be specified in order for any run-time added custom converters to be found and to use `PropertyNameCaseInsensitive` for property lookup with `JsonObject`.
+
+It is cumbersome and error-prone to specify the options instance for every new element or property:
+```cs
+    var jObject = new JsonObject(options)
+    {
+        ["MyString"] = new JsonValue("Hello!", options),
+        ["MyBoolean"] = new JsonValue(false, options),
+        ["MyArray"] = new JsonArray(options)
+        {
+            new JsonValue(2, options),
+            new JsonValue(3, options),
+            new JsonValue(42, options)
+        },
+    }
+```
+so it is possible to omit the options instance for non-root members. When a property or element is added, the options instance is set to the parent's instance:
```

Comment:
How would this technically work? Do JsonToken instances have a reference to their parent? If so, what happens when a JsonObject instance is added as the child of two different parents?

In Newtonsoft.Json the `JToken.Parent` property works because adding a token that already has a parent to a different parent will clone it.